Paging wip

elliott-beach opened the pull request:
@kstowe Checkout out the page fault handler in `main.c`.
It's not working yet, and the final code will probably look a lot different, but this is what I imagine
for a minimum viable example.

https://github.com/elliott-beach/project2_base/blob/12f7f5e94b6a12e14801c078ffaa1828ff9c95ef/main.c#L20-L56

elliott-beach committed:
first attempt at random paging

elliott-beach committed:
test out method behavior

elliott-beach committed:
paging - WIP

kstowe commented on main.c:
```diff
@@ -15,14 +15,47 @@ how to use the page table and disk interfaces.
+		int page;
+		for (page=0; page<pt->npages; page++){
+			if(pt->page_mapping[page] == f){ /* && page_bits[page] != 0 */ // We could also check the page bits, but it doesn't matter too much.
+				break;
```
I think we need to check page_bits[page] too because initially all the page_mappings are going to be equal to 0 so technically every page is mapped to the first frame.

ghost commented:
Yeah that looks pretty good. All we have to do is tweak the evict page part to implement each of the algorithms.

ghost commented:
Do you think we could use a static variable in the handler to keep track of how many page frames have been allocated? This would make the empty case simpler and it would make FIFO pretty easy to implement

elliott-beach commented:
Thanks for the feedback, I made some of those changes and fixed some problems and have been debugging it for a while :P, must push.

Your suggestion about checking the page bits is good. That fixes a problem I’ve noticed.

elliott-beach commented:
Yea the static variable would be a much better way of doing things, the current code is something like O(n^2) with number of frames, too.

elliott-beach committed:
WIP: Code seems to be going in infinite loop.

some changes:
1. use page#, not frame#, for disk io
2. rename loop vars to be readable
3. fix off-by-one err in eviction condition
4. fix loop condition for finding the first page in memory
etc.

elliott-beach commented:
@kstowe I pushed my code again, but still can't get it to work. It goes into an infinite loop, could you see if you could fiddle with it to see what's wrong?

elliott-beach committed:
got random paging to work!

elliott-beach committed:
print output formatting

elliott-beach commented:
So now I actually figured it out, I think. The problem was not that the code was trapped in an endless loop, just it was alternating between 2 pages, and with printing to stdout on each page fault it would take approximately infinity seconds to complete. 

Changing it to random paging fixed the problem.

Also, I bet when you added the `&& page_bits[page] != 0` check you used `page`, not `curr_page`, this would cause the program to leave the endless loop, even though it is not correct (should be `curr_page`).
